pkg/chain/ethereum/ethereum.go

```diff
@@ -18,17 +22,75 @@ var logger = log.Logger("keep-chain-ethereum")
 // provides the implementation of generic features like balance monitor,
 // block counter and similar.
 type Chain struct {
-	key          *keystore.Key
-	client       ethutil.EthereumClient
+	key     *keystore.Key
+	client  ethutil.EthereumClient
+	chainID *big.Int
+
 	blockCounter *ethlike.BlockCounter
+	nonceManager *ethlike.NonceManager
+	miningWaiter *ethutil.MiningWaiter
+
+	// transactionMutex allows interested parties to forcibly serialize
+	// transaction submission.
+	//
+	// When transactions are submitted, they require a valid nonce. The nonce is
+	// equal to the count of transactions the account has submitted so far, and
+	// for a transaction to be accepted it should be monotonically greater than
+	// any previous submitted transaction. To do this, transaction submission
+	// asks the Ethereum client it is connected to for the next pending nonce,
+	// and uses that value for the transaction. Unfortunately, if multiple
+	// transactions are submitted in short order, they may all get the same
+	// nonce. Serializing submission ensures that each nonce is requested after
+	// a previous transaction has been submitted.
+	transactionMutex *sync.Mutex
 }
 
-// NewChain construct a new instance of the Ethereum chain handle.
-func NewChain(
+// Connect creates Random Beacon and TBTC Ethereum chain handles.
+func Connect(
 	ctx context.Context,
-	config *ethereum.Config,
-	client ethutil.EthereumClient,
+	config ethereum.Config,
```

Comment:
Any particular reason to pass the config as a value and not a pointer?

Reply:
My reasoning was that we do not want to modify anything in the original config and that the value passed should be read-only. This works better if we pass by value (a copy) instead of a pointer.